uefi/src/fs/file_system.rs

```diff
@@ -0,0 +1,312 @@
+//! Module for [FileSystem].
+
+use super::*;
+use super::dir_entry_iter::UefiDirectoryIter;
+use super::normalized_path::{NormalizedPath, PathError};
+use super::path::Path;
+use alloc::boxed::Box;
+use alloc::string::{FromUtf8Error, String, ToString};
+use alloc::vec::Vec;
+use core::fmt;
+use core::fmt::{Debug, Formatter};
+use core::ops::Deref;
+use derive_more::Display;
+use uefi::proto::media::file::{FileAttribute, FileInfo, FileType};
+use uefi::table::boot::ScopedProtocol;
+use alloc::format;
+use log::info;
+use uefi::fs::path::Component;
+use crate::fs::path::Components;
+
+/// All errors that can happen when working with the [`FileSystem`].
+#[derive(Debug, Clone, Display, PartialEq, Eq)]
+pub enum FileSystemError {
+    /// Can't open the root directory of the underlying volume.
+    CantOpenVolume,
+    /// The path is invalid because of the underlying [`PathError`].
+    IllegalPath(PathError),
+    /// The file or directory was not found in the underlying volume.
+    FileNotFound(String),
+    /// The path is existend but does not correspond to a directory when a directory was expected.
+    NotADirectory(String),
+    /// The path is existend but does not correspond to a file when a file was expected.
+    NotAFile(String),
+    /// Can't delete the file.
+    CantDeleteFile(String),
+    /// Can't delete the directory.
+    CantDeleteDirectory(String),
+    /// Error writing bytes.
+    WriteFailure,
+    /// Error flushing file.
+    FlushFailure,
+    /// Error reading file.
+    ReadFailure,
+    /// Can't parse file content as UTF-8.
+    Utf8Error(FromUtf8Error),
+    /// Could not open the given path.
+    OpenError(String),
+}
+
+#[cfg(feature = "unstable")]
+impl core::error::Error for FileSystemError {}
+
+impl From<PathError> for FileSystemError {
+    fn from(err: PathError) -> Self {
+        Self::IllegalPath(err)
+    }
+}
+
+/// Return type for public [`FileSystem`] operations.
+pub type FileSystemResult<T> = Result<T, FileSystemError>;
+
+/// High-level file-system abstraction for UEFI volumes with an API that is
+/// close to `std::fs`. It acts as convenient accessor around the
+/// [`SimpleFileSystemProtocol`].
+pub struct FileSystem<'boot_services> {
+    /// Underlying UEFI protocol.
+    proto: ScopedProtocol<'boot_services, SimpleFileSystemProtocol>,
+}
```

Comment:
I'd prefer to keep thinks separate. Low-level API and high-level API should not exist on the same type, especially, as mixing them together might break stuff.

Reply:
Can you say more about what might break when using both? If there are things the end user must avoid doing then we should add that info to the docs.

Reply:
Correction, I mixed up the SimpleFileSystem protocol with the FileProtocol. As the SimpleFileSystem protocol is so simple, it only has `open_volume`, integrating the new FileSystem abstraction into the SimpleFileSystem protocol abstraction is a possible design choice. However, the following properties do not feel right:

- mix of low-level API with high-level API
- I need additional helpers for the Filesystem (`Path`, `FileSystemError`,  ...) and it feels wrong to integrate them into `uefi::proto::media`. They are types of a different level.

Instead, I'd allow constructing a `FileSystem` from a `SimpleFileSystem` (via .into() or `.to_file_ystem()`) for a high developer convenience.

What do you say?

Reply:
That makes sense, I especially agree that it would feel a little weird to have things like `Path` mixed into `uefi::proto::media`. It's a little unfortunate that the names are kinda reversed; `SimpleFileSystem` is really more complex to use than `FileSystem`. But that's probably fine, I don't have any better suggestion :) Anyway, +1 to the separate module.